public/app/features/dashboard/dashgrid/liveTimer.ts

```diff
@@ -0,0 +1,113 @@
+import { dateMath, dateTime, TimeRange } from '@grafana/data';
+import { PanelChrome } from './PanelChrome';
+
+// target is 20hz (50ms), but we poll at 100ms to smooth out jitter
+const interval = 100;
+
+interface LiveListener {
+  last: number;
+  intervalMs: number;
+  panel: PanelChrome;
+}
+
+class LiveTimer {
+  listeners: LiveListener[] = [];
+
+  budget = 1;
+  threshold = 1.5; // trial and error appears about right
+  ok = true;
+  lastUpdate = Date.now();
+
+  isLive = false; // the dashboard time range ends in "now"
+  timeRange?: TimeRange;
+
+  /** Called when the dashboard time range changes */
+  setLiveTimeRange(v?: TimeRange) {
+    this.timeRange = v;
+    this.isLive = v?.raw?.to === 'now';
+
+    if (this.isLive) {
+      for (const listener of this.listeners) {
+        listener.intervalMs = getLiveTimerInterval(v!, listener.panel.props.width);
+      }
+    }
+  }
+
+  listen(panel: PanelChrome) {
+    this.listeners.push({
+      last: this.lastUpdate,
+      panel: panel,
+      intervalMs: getLiveTimerInterval(
+        this.timeRange ?? ({ from: dateTime(0), to: dateTime(1000), raw: { from: '', to: '' } } as TimeRange),
+        panel.props.width
+      ),
+    });
+  }
+
+  remove(panel: PanelChrome) {
+    this.listeners = this.listeners.filter((v) => v.panel !== panel);
```

Comment:
maybe use Set for `this.listeners`, that way this can be .delete(), and you get free dedupes during .add()

Reply:
maybe... remove is called super infrequently (when navigating from one page to the next), but iterating over the listeners is called all the time.   I think it would need to be a Map<panel,listener> and call Object.values() in each loop.  Your thoughts?